tensorflow/contrib/tensorrt/kernels/trt_calib_op.cc

```diff
@@ -0,0 +1,134 @@
+        LOG(FATAL) << "Unsupported Data type "                 \
+                   << tensorflow::DataTypeString(TENSOR_TYPE); \
+        break;                                                 \
+      }                                                        \
+    }                                                          \
+  }
+
+void TRTCalibOp::Compute(tensorflow::OpKernelContext* ctx) {
+  auto trt_rm = tensorflow::trt::TRTResourceManager::instance();
+  auto resmgr = trt_rm->getManager("TRTCalibOps");
+  tensorflow::trt::TRTCalibrationResource* calibRes = nullptr;
+  auto status = resmgr->Lookup(resource_name_, resource_name_, &calibRes);
+
+  if (!status.ok()) {
+    ctx->SetStatus(status);
+    return;
+  }
+  int numInputs = ctx->num_inputs();
+  if (calibRes->calibrator_ == nullptr) {  // first run instantiate calibrator
+    dev_tensors_.resize(numInputs);
+    int batchSize = ctx->input(0).dim_size(0);
+    VLOG(1) << " Constructing calibrator";
+    for (int i = 0; i < numInputs;
+         i++) {  // allocate workspace on device for inputs
+      const tensorflow::Tensor& t = ctx->input(i);
+      OP_REQUIRES_OK(ctx,
+                     ctx->allocate_persistent(t.dtype(), t.shape(),
+                                              &dev_tensors_.at(i), nullptr));
+      const auto dTensor = dev_tensors_.at(i).AccessTensor(ctx);
+      CHECK_EQ(t.TotalBytes(), dTensor->TotalBytes());
+      void* devAddr = nullptr;
+      GET_TENSOR_ADDRESS(dTensor, devAddr);
+      device_buffers_.emplace(
+          input_names_.at(i),
+          std::pair<void*, size_t>(devAddr, dTensor->TotalBytes()));
+    }
+
+    calibRes->calibrator_ =
+        new TRTInt8Calibrator(device_buffers_, batchSize, resource_name_);
+    string label(resource_name_);
+    calibRes->thr_ = new std::thread([calibRes, label]() {
+      VLOG(1) << "Starting calibration thread, Calibration Resource @ "
+              << calibRes;
+      calibRes->builder_->setInt8Calibrator(calibRes->calibrator_);
+      calibRes->builder_->setInt8Mode(true);
+      calibRes->engine_ = calibRes->builder_->buildCudaEngine(
+          *calibRes->network_);  // will loop until we terminate calibrator
+      VLOG(1) << "Calibration loop terminated " << label;
+    });
+    VLOG(1) << "initialized calibrator resource";
+  }  //  calibrator initialized
+
+  // Pass input data to calibrator
+  std::unordered_map<string, void*> input_data;
+  for (int i = 0; i < numInputs; i++) {
+    const Tensor& t = ctx->input(i);
+    void* data_address = nullptr;
+    const Tensor* t_ptr = &t;
+    GET_TENSOR_ADDRESS(t_ptr, data_address);
+    const auto dTensor = dev_tensors_.at(i).AccessTensor(ctx);
+    CHECK_EQ(t.TotalBytes(),
+             dTensor->TotalBytes());  // use the tensor so FW keeps it
+    input_data.emplace(input_names_.at(i), data_address);
+    ctx->set_output(i, t);
+  }
+  VLOG(2) << "Filled map for sending";
+  calibRes->calibrator_->setBatch(input_data);
```

Comment:
I think we need to wait for the batch to be consumed before returning? Maybe adding a method named e.g. WaitForCalibration() in the calibrator and invoke it here?

Reply:
setBatch() returns when it copies contents to workspace buffers. It will block until buffer is consumed. WaitForCalibration will block the execution of the graph which is not necessary for the calibration. When I convert this to async op, it will be a similar approach.

Reply:
Ok, I think this is fine when in calibration mode. A related question is, when does it call setDone()? And when it call setDone() it should wait until all batches are consumed and calibration is finished, right? If that would be in a future PR please add a TODO.

Reply:
Since we don't have any way to notify the end of calibration, it is done in another step when user converts the calibration graph to inference graph. That code is part of a bigger PR that is going to come after this one. We split the PR into two to reduce the number of lines in the bigger PR.